File: linkedList/cppRealizations/linkedListCycle.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        while(head){
            if(head->val == INT_MAX){
                return true;
            }
            head->val = INT_MAX;
            head = head->next;
        }
        return false;
    }
};


int main(){
    vector<ListNode*> input = {new ListNode(3), new ListNode(2), new ListNode(0), new ListNode(-4)};
    input[0]->next = input[1];
    input[1]->next = input[2];
    input[2]->next = input[3];
    input[3]->next = input[1];
    
    
    cout << Solution().hasCycle(input[0]) << endl;
    return 0;
}

File: linkedList/linkedListCycle.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def hasCycle(self, head):
        while head:
            if head.val == float("inf"):
                return True
            head.val = float("inf")
            head = head.next
        return False


if __name__ == "__main__":
    temp = [ListNode(3), ListNode(2), ListNode(0), ListNode(-4)]
    temp[0].next = temp[1]
    temp[1].next = temp[2]
    temp[2].next = temp[3]
    temp[3].next = temp[1]

    start = temp[0]

    print("Input:")
    for _ in range(5):
        print(start.val)
        start = start.next

    print("Solution:")
    start = temp[0]
    print(Solution().hasCycle(start))